### ISSUE TITLE
Prevent Ingress Rule Creation

### ISSUE THREAD

**Issue by ambsw-technology** [cfn-modules/alb-listener#7]
We'd like to be able to separate CF code that deploys application architecture from CF code that deploys ingress networking architecture.

We're going to be giving our software to a partner to run on their own account.  We'd like to be sure the application deployment architecture is exactly the same even if they use a different ingress strategy.  I imagine there are other cases where you'd want to configure the networking after the application is setup including e.g.

 - Giving access to QA/configuration teams before switching an app to a client or the public internet
 - Configuring an instance before establishing a VPC Peer with a client (the peer must be `active` to actually use the remote SG in an ingress rule).

In the current codebase,

 - The ALB Target Group is "optional" in e.g. a Fargate Service, but "You cannot add, remove, or change the load balancer configuration of an existing service" ([docs](https://docs.aws.amazon.com/AmazonECS/latest/developerguide/service-load-balancing.html#load-balancing-considerations)) so it's actually required if you want to use the feature.
 - The ALB Target Group requires an ALB Listener**
 - The ALB Listener creates one of two Ingress rules (from an SG or `0.0.0.0/0`).

** The `TargetGroup` in [`ecs-alb-target`](https://github.com/cfn-modules/ecs-alb-target/blob/master/module.yml) doesn't depend on the `ListenerRule` so the dependency could be inverted.

You can "work around" this default by providing an unused SG to the template, but I suspect this would be a red flag during a security review.  Otherwise, you're left with `0.0.0.0/0` which violates least privilege in almost every case (except for an internet-facing system).

My best idea so far is to add a parameter that controls the `0.0.0.0/0` rule e.g. `CreateWorldIngress` that defaults to "conditional" (the current behavior) but can be set to "true" or "false".  I'd rather it default to "false" but understand the potential issues with backwards compatibility.

**Comment by michaelwittig** [comment]
I'm not sure if I get this right:

> The ALB Target Group is "optional" in e.g. a Fargate Service, but "You cannot add, remove, or change the load balancer configuration of an existing service" (docs) so it's actually required if you want to use the feature.

Do you think about changing the ECS service outside of CloudFormation after creation?

---

One possible solution I can think of: We add a parameter to the `alb-listener` module to control the `CidrIp` value. The parameter value defaults to `0.0.0.0/0`. If the parameter value is empty we don't create a rule at all. Would that work for you?

**Comment by ambsw-technology** [comment]
> Do you think about changing the ECS service outside of CloudFormation after creation?

No.  I was just making the point that you cannot add the ALB to the service after it's created.

I'd prefer to create a security group with no rules when I create the ALB.  If the Security Group was exposed (interface), I could come back and create the rules later as part of my ingress template.

A `CidrIp` or `CidrRange` may be helpful to someone, but it's no more secure (and probably less) than `ClientSgModule` for our use case:

 - With `ClientSgModule` I can permit a Security Group that has no devices in it.  There's literally no way for a device to use the SG unless it's intentionally attached.  A human could do this by mistake, but they'd realize that it was a mistake.
 - A `CidrRange` implicitly gives an IP address/range access.  It's at least theoretically possible that someone would intentionally VPC Peer/VPN and route that IP address or range (e.g. we establish VPNs with lots of clients who use lots of IP ranges/addresses).  That admin would have no idea (and no reason to know) that they were giving a random device access to the ALB.

**Comment by michaelwittig** [comment]
You can add an ALB to the service. CloudFormation will just replace the service with a new one.

---

I'm not sure if I get you. I suggested that "If the parameter value is empty we don't create a rule at al". So you could manage you own rules without a default rule?

**Comment by ambsw-technology** [comment]
Yes.  My layer cake of templates include:

    Common Resources like Alerting
        Environment Resources like the VPC and ALB
            Application-level Resources like DB and ECS
            Ingress networking

The idea is to have a stable environment that can accept multiple Application and Ingress Networking stacks.  For example, I can start with a WorkSpace for admin and a VPN for client access.  If I later decide to switch to a client-to-site VPN for admin and a PrivateLink for client access, I replace those templates.  I don't have to go back and touch the ALB (or Application stacks).

**Comment by ambsw-technology** [comment]
Yes.  My layer cake of templates include:

    Common Resources like Alerting
        Environment Resources like the VPC and ALB
            Application-level Resources like DB and ECS
            Ingress networking

The idea is to have a stable environment that can accept multiple Application and Ingress Networking stacks.  For example, I can start with a WorkSpace for admin and a VPN for client access.  If I later decide to switch to a client-to-site VPN for admin and a PrivateLink for client access, I replace those templates.  I don't have to go back and touch the ALB (or Application stacks).
